packages/transport-webrtc/test/certificate-persistent.spec.ts

```diff
@@ -0,0 +1,176 @@
+    sandbox.restore()
+  })
+  
+  describe('certificate integration', function () {
+    // These tests might take longer to run
+    this.timeout(20000)
+
+    // Helper function to create a node with keychain
+    async function createNode(persistentDatastore: MemoryDatastore | null = null, options = {}) {
+      const nodeDatastore = persistentDatastore || new MemoryDatastore()
+      
+      return await createLibp2p({
+        addresses: {
+          listen: ['/ip4/127.0.0.1/udp/0/webrtc-direct']
+        },
+        datastore: nodeDatastore,
+        connectionGater: {
+          denyDialMultiaddr: async () => false
+        },
+        streamMuxers: [mplex()],
+        connectionEncrypters: [noise()],
+        transports: [
+          webRTCDirect({
+            ...options,
+            dataChannel: {
+              maxMessageSize: 1 << 16
+            }
+          })
+        ],
+        services: {
+          keychain: async (components) => {
+            const keychainModule = await import('@libp2p/keychain')
+            return keychainModule.keychain({
+              pass: 'very-secure-password-for-testing',
+              dek: {
+                keyLength: 512 / 8,
+                iterationCount: 1000,
+                salt: 'at-least-16-characters-long-for-testing',
+                hash: 'sha2-512'
+              }
+            })(components)
+          }
+        }
+      })
+    }
+    
+    it('should use certificate hash in multiaddr', async () => {
+      // Create a node with certificate persistence enabled
+      const node = await createNode(null, {
+        certificates: [], // Empty array to not use any pre-defined certificates
+        useLibjuice: false, // Use WebRTC's built-in STUN/TURN
+        rtcConfiguration: {
+          iceTransportPolicy: 'all',
+        }
+      })
+      
+      try {
+        await node.start()
+
+        // Wait for addresses to be available
+        await new Promise(resolve => setTimeout(resolve, 1000))
+        
+        // Check that at least one address includes the webrtc-direct protocol
+        const addrs = node.getMultiaddrs()
+        const webrtcAddrs = addrs.filter(ma => ma.toString().includes('/webrtc-direct'))
+        
+        expect(webrtcAddrs.length).to.be.greaterThan(0)
+        
+        // Check that the address includes a certhash component
+        const addrWithCertHash = webrtcAddrs.find(ma => ma.toString().includes('/certhash/'))
+        expect(addrWithCertHash).to.exist
+        
+        testLogger('Found WebRTC address with certificate hash: %s', addrWithCertHash?.toString())
+      } finally {
+        await node.stop()
+      }
+    })
+    
+    /**
+     * This test checks that the WebRTC transport is at least attempting to persist and retrieve certificates.
+     * After fixing the createPrivateKeyFromCertificate implementation, it should now work properly.
+     */
+    it('should attempt certificate persistence between restarts', async function() {
```

Comment:
This test should just start a node, record any WebRTC-Direct multiaddrs it's listening on, then stop the node and start it again and assert that it's re-used the same cert.  No need for deep stubbing of module internals.